File: tools/lint_definitions.py
#!/usr/bin/env python3
"""Validate task-definition files against the Swale platform contract.

Run locally:  python tools/lint_definitions.py
CI runs this on every pull request.
"""
from __future__ import annotations

import glob
import os
import re
import sys

import yaml

NAME_RE = re.compile(r"^[A-Za-z0-9_-]{1,100}$")
IDENT_RE = re.compile(r"^[a-z0-9_]+$")  # snake_case identifiers
DIGEST_RE = re.compile(r"@sha256:[0-9a-f]{64}$")

TOP_KEYS = {"name", "description", "inputs", "outputs", "exec"}
EXEC_KEYS = {"image", "args"}
INPUT_KEYS = {"description", "required", "default"}
OUTPUT_KEYS = {"description"}


def split_frontmatter(text: str):
    if not text.startswith("---"):
        return None, None
    end = text.find("\n---", 3)
    if end == -1:
        return None, None
    fm = text[3:end].lstrip("\n")
    body = text[end + 4 :].lstrip("\n\r")
    return fm, body


def lint_file(path: str) -> list[str]:
    errors: list[str] = []
    rel = os.path.relpath(path).replace("\\", "/")

    parts = rel.split("/")
    if len(parts) != 3 or parts[0] != "tasks" or parts[2] != "task.md":
        errors.append(f"{rel}: must live at tasks/<name>/task.md (versions are git tags '<name>/<version>')")
        return errors
    name = parts[1]

    if not NAME_RE.match(name):
        errors.append(f"{rel}: task name '{name}' must match {NAME_RE.pattern} (no dots)")

    with open(path, "r", encoding="utf-8") as handle:
        text = handle.read()

    if len(text.encode("utf-8")) > 1024 * 1024:
        errors.append(f"{rel}: file exceeds 1 MiB")

    fm, body = split_frontmatter(text)
    if fm is None:
        errors.append(f"{rel}: missing YAML frontmatter starting at byte 0")
        return errors
    if not body:
        errors.append(f"{rel}: body must be non-empty")

    try:
        data = yaml.safe_load(fm)
    except yaml.YAMLError as exc:
        errors.append(f"{rel}: frontmatter is not valid YAML: {exc}")
        return errors
    if not isinstance(data, dict):
        errors.append(f"{rel}: frontmatter must be a mapping")
        return errors

    unknown = set(data) - TOP_KEYS
    if unknown:
        errors.append(f"{rel}: unknown top-level keys: {sorted(unknown)}")

    exec_block = data.get("exec")
    if not isinstance(exec_block, dict):
        errors.append(f"{rel}: 'exec' is required and must be a mapping")
    else:
        unknown_exec = set(exec_block) - EXEC_KEYS
        if unknown_exec:
            errors.append(f"{rel}: unknown exec keys: {sorted(unknown_exec)}")
        image = exec_block.get("image")
        if not isinstance(image, str) or not image.strip():
            errors.append(f"{rel}: exec.image is required and must be a non-empty string")
        elif not DIGEST_RE.search(image):
            errors.append(f"{rel}: exec.image must be digest-pinned (…@sha256:<64 hex>): '{image}'")
        args = exec_block.get("args")
        if args is not None and not isinstance(args, list):
            errors.append(f"{rel}: exec.args must be a list of strings")

    for section, allowed in (("inputs", INPUT_KEYS), ("outputs", OUTPUT_KEYS)):
        block = data.get(section)
        if block is None:
            continue
        if not isinstance(block, dict):
            errors.append(f"{rel}: '{section}' must be a mapping")
            continue
        for ident, spec in block.items():
            if not IDENT_RE.match(str(ident)):
                errors.append(f"{rel}: {section} identifier '{ident}' must be snake_case ({IDENT_RE.pattern})")
            if spec is not None and not isinstance(spec, dict):
                errors.append(f"{rel}: {section}.{ident} must be a mapping or empty")
                continue
            if isinstance(spec, dict):
                unknown_field = set(spec) - allowed
                if unknown_field:
                    errors.append(f"{rel}: {section}.{ident} has unknown keys: {sorted(unknown_field)}")

    return errors


def main() -> int:
    root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    files = sorted(glob.glob(os.path.join(root, "tasks", "**", "*.md"), recursive=True))
    if not files:
        print("No task definitions found under tasks/.")
        return 0

    all_errors: list[str] = []
    for path in files:
        all_errors.extend(lint_file(path))

    if all_errors:
        for error in all_errors:
            print(f"::error::{error}")
        print(f"\n{len(all_errors)} problem(s) in {len(files)} file(s).")
        return 1

    print(f"OK: {len(files)} task definition(s) valid.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/lint-definitions.mjs
#!/usr/bin/env node
// Validate task-definition files against the Swale platform contract.
//
// Run locally:  npm install && npm run lint
// CI runs this on every pull request and on every release tag.

import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, dirname, relative, sep } from "node:path";
import { fileURLToPath } from "node:url";
import yaml from "js-yaml";

const NAME_RE = /^[A-Za-z0-9_-]{1,100}$/;
const IDENT_RE = /^[a-z0-9_]+$/; // snake_case identifiers
const DIGEST_RE = /@sha256:[0-9a-f]{64}$/;
const ENV_NAME_RE = /^[A-Za-z_][A-Za-z0-9_]*$/;
const ENV_RESERVED_PREFIXES = ["WORKFLOW_", "INPUT_"];
const DESCRIPTION_MAX_LENGTH = 100;
const MAX_FILE_BYTES = 1024 * 1024;

const TOP_KEYS = new Set(["name", "description", "inputs", "outputs", "exec"]);
const EXEC_KEYS = new Set(["image", "args", "env"]);
const INPUT_KEYS = new Set(["description", "required", "default"]);
const OUTPUT_KEYS = new Set(["description"]);

const isMapping = (value) =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const unknownKeys = (value, allowed) =>
  Object.keys(value)
    .filter((key) => !allowed.has(key))
    .sort();

function splitFrontmatter(text) {
  if (!text.startsWith("---")) {
    return [null, null];
  }
  const end = text.indexOf("\n---", 3);
  if (end === -1) {
    return [null, null];
  }
  const frontmatter = text.slice(3, end).replace(/^\n+/, "");
  const body = text.slice(end + 4).replace(/^[\n\r]+/, "");
  return [frontmatter, body];
}

function lintFile(root, path) {
  const errors = [];
  const rel = relative(root, path).split(sep).join("/");

  const parts = rel.split("/");
  if (parts.length !== 3 || parts[0] !== "tasks" || parts[2] !== "task.md") {
    errors.push(`${rel}: must live at tasks/<name>/task.md (versions are git tags '<name>/<version>')`);
    return errors;
  }
  const name = parts[1];

  if (!NAME_RE.test(name)) {
    errors.push(`${rel}: task name '${name}' must match ${NAME_RE.source} (no dots)`);
  }

  const text = readFileSync(path, "utf8");
  if (Buffer.byteLength(text, "utf8") > MAX_FILE_BYTES) {
    errors.push(`${rel}: file exceeds 1 MiB`);
  }

  const [frontmatter, body] = splitFrontmatter(text);
  if (frontmatter === null) {
    errors.push(`${rel}: missing YAML frontmatter starting at byte 0`);
    return errors;
  }
  if (!body) {
    errors.push(`${rel}: body must be non-empty`);
  }

  let data;
  try {
    data = yaml.load(frontmatter);
  } catch (error) {
    errors.push(`${rel}: frontmatter is not valid YAML: ${error.message}`);
    return errors;
  }
  if (!isMapping(data)) {
    errors.push(`${rel}: frontmatter must be a mapping`);
    return errors;
  }

  const unknownTop = unknownKeys(data, TOP_KEYS);
  if (unknownTop.length > 0) {
    errors.push(`${rel}: unknown top-level keys: ${JSON.stringify(unknownTop)}`);
  }

  const description = data.description;
  if (typeof description !== "string" || description.trim() === "") {
    errors.push(`${rel}: description is required and must be a non-empty string`);
  } else if (description.length > DESCRIPTION_MAX_LENGTH) {
    errors.push(
      `${rel}: description must be at most ${DESCRIPTION_MAX_LENGTH} characters ` +
        `(is ${description.length}) — the platform rejects longer descriptions on import`,
    );
  }

  const execBlock = data.exec;
  if (!isMapping(execBlock)) {
    errors.push(`${rel}: 'exec' is required and must be a mapping`);
  } else {
    const unknownExec = unknownKeys(execBlock, EXEC_KEYS);
    if (unknownExec.length > 0) {
      errors.push(`${rel}: unknown exec keys: ${JSON.stringify(unknownExec)}`);
    }

    const image = execBlock.image;
    if (typeof image !== "string" || image.trim() === "") {
      errors.push(`${rel}: exec.image is required and must be a non-empty string`);
    } else if (!DIGEST_RE.test(image)) {
      errors.push(`${rel}: exec.image must be digest-pinned (…@sha256:<64 hex>): '${image}'`);
    }

    if (execBlock.args !== undefined && !Array.isArray(execBlock.args)) {
      errors.push(`${rel}: exec.args must be a list of strings`);
    }

    const env = execBlock.env;
    if (env !== undefined) {
      if (!isMapping(env)) {
        errors.push(`${rel}: exec.env must be a mapping`);
      } else {
        for (const varName of Object.keys(env)) {
          if (!ENV_NAME_RE.test(varName)) {
            errors.push(`${rel}: exec.env name '${varName}' must match ${ENV_NAME_RE.source}`);
          }
          if (ENV_RESERVED_PREFIXES.some((prefix) => varName.startsWith(prefix))) {
            errors.push(`${rel}: exec.env name '${varName}' uses a reserved prefix (WORKFLOW_/INPUT_)`);
          }
        }
      }
    }
  }

  for (const [section, allowed] of [
    ["inputs", INPUT_KEYS],
    ["outputs", OUTPUT_KEYS],
  ]) {
    const block = data[section];
    if (block === undefined || block === null) {
      continue;
    }
    if (!isMapping(block)) {
      errors.push(`${rel}: '${section}' must be a mapping`);
      continue;
    }
    for (const [ident, spec] of Object.entries(block)) {
      if (!IDENT_RE.test(ident)) {
        errors.push(`${rel}: ${section} identifier '${ident}' must be snake_case (${IDENT_RE.source})`);
      }
      if (spec !== null && spec !== undefined && !isMapping(spec)) {
        errors.push(`${rel}: ${section}.${ident} must be a mapping or empty`);
        continue;
      }
      if (isMapping(spec)) {
        const unknownField = unknownKeys(spec, allowed);
        if (unknownField.length > 0) {
          errors.push(`${rel}: ${section}.${ident} has unknown keys: ${JSON.stringify(unknownField)}`);
        }
      }
    }
  }

  return errors;
}

function findMarkdown(directory) {
  const found = [];
  for (const entry of readdirSync(directory)) {
    const full = join(directory, entry);
    if (statSync(full).isDirectory()) {
      found.push(...findMarkdown(full));
    } else if (entry.endsWith(".md")) {
      found.push(full);
    }
  }
  return found.sort();
}

const root = dirname(dirname(fileURLToPath(import.meta.url)));
const files = findMarkdown(join(root, "tasks"));

if (files.length === 0) {
  console.log("No task definitions found under tasks/.");
  process.exit(0);
}

const allErrors = files.flatMap((path) => lintFile(root, path));

if (allErrors.length > 0) {
  for (const error of allErrors) {
    console.log(`::error::${error}`);
  }
  console.log(`\n${allErrors.length} problem(s) in ${files.length} file(s).`);
  process.exit(1);
}

console.log(`OK: ${files.length} task definition(s) valid.`);
